### PR TITLE
Made setting of proxied headers deterministic based on configuration alone

### PR DESCRIPTION
**Pull request by aeijdenberg**
## Description

Previously some headers that are normally set by the proxy (and may be
replied upstream for authorization decisiions) were not being set
depending on values in the users session.

This change ensure that if a given header is sometimes set, it will
always be either set or removed.

It might be worth considerating always deleting these headers if we
didn't add them.

## Motivation and Context

Stop malicious users being able to set headers that might be used for authorization.

## How Has This Been Tested?

Existing unit tests aren't broken.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by aeijdenberg** [comment]
Have added entry to the changelog as requested (and fixed some URL copy/paste typos).

File: oauthproxy.go [hunk 1]
```diff
@@ -810,19 +810,30 @@ func (p *OAuthProxy) getAuthenticatedSession(rw http.ResponseWriter, req *http.R
 // addHeadersForProxying adds the appropriate headers the request / response for proxying
 func (p *OAuthProxy) addHeadersForProxying(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) {
 	if p.PassBasicAuth {
+		req.Header.Del("Authorization")
+		req.Header.Del("X-Forwarded-Email")
+		req.Header.Del("X-Forwarded-User")
```
Comment (steakunderscore) [inline oauthproxy.go]: Why do we need to delete the `X-Forwarded-User` header when it's always set (on line 818)? Basically what's the benefit?

File: oauthproxy.go [hunk 2]
```diff
@@ -810,19 +810,30 @@ func (p *OAuthProxy) getAuthenticatedSession(rw http.ResponseWriter, req *http.R
...
+
 		req.SetBasicAuth(session.User, p.BasicAuthPassword)
 		req.Header["X-Forwarded-User"] = []string{session.User}
 		if session.Email != "" {
 			req.Header["X-Forwarded-Email"] = []string{session.Email}
 		}
```
Comment (steakunderscore) [inline oauthproxy.go]: Could setting the `X-Forwarded-Email` header be tidied up with the following?
```
if session.Email != "" {
  req.Header["X-Forwarded-Email"] = []string{session.Email}
} else {
  req.Header.Del("X-Forwarded-Email")
}
```

File: oauthproxy.go (hunk 1 above)
Comment (aeijdenberg) [reply oauthproxy.go]: The intent was simply to make each section the same - start with removing any headers that it purports to set, then set (sometimes conditionally, and sometimes, like this one, always) various values.

File: oauthproxy.go (hunk 2 above)
Comment (aeijdenberg) [reply oauthproxy.go]: Again same answer as the other one. Yes each section could individually be technically simpler, but it felt more robust to do the same boring thing each time, ie delete all old headers then perform the logic nessary to add new ones.

File: oauthproxy.go [hunk 3]
```diff
@@ -810,19 +810,30 @@ func (p *OAuthProxy) getAuthenticatedSession(rw http.ResponseWriter, req *http.R
...
+
 		req.Header["X-Forwarded-User"] = []string{session.User}
 		if session.Email != "" {
 			req.Header["X-Forwarded-Email"] = []string{session.Email}
 		}
 	}
 	if p.SetXAuthRequest {
+		rw.Header().Del("X-Auth-Request-Access-Token")
+		rw.Header().Del("X-Auth-Request-Email")
+		rw.Header().Del("X-Auth-Request-User")
```
Comment (ploxiln) [inline oauthproxy.go]: This runs before `p.serveMux.ServeHTTP(rw, req)`, so I think it is useless to try delete headers of the response writer here. If they are set by the upstream, they will happen *after* this.

File: oauthproxy.go (hunk 2 above)
Comment (JoelSpeed) [reply oauthproxy.go]: I think from a readability perspective, changing to `if { .. } else { .. } ` actually would make a lot of sense. The logic we want is that we set the header or remove it if we aren't setting it, this is exactly encoded into if/else and would make the code easier to reason with in the future. By doing this we could also get rid of the deletion of headers that are always set and the flow would still be consistent throughout, WDYT @aeijdenberg ?

File: oauthproxy.go (hunk 3 above)
Comment (JoelSpeed) [reply oauthproxy.go]: This is a very good point from @ploxiln, if we make the change to if/else statements, we can drop these Headers from the response as they _should_ at this point be empty anyway

**Comment by steakunderscore** [comment]
@aeijdenberg Do you think you'll have time to fix this up? It'd be great to get it merged so we can cut a new release.

**Comment by aeijdenberg** [comment]
@steakunderscore @JoelSpeed - I've made changes that I think reflect the style of code that you're looking for here. I tend to disagree with much of the feedback - personally I now find this code harder to reason with than the earlier PR, but we need to keep moving forward, so hopefully these changes match better what you're looking for.

I've moved roles and no longer have an active interest in this code so I won't have much time to do many more rounds on it.

File: oauthproxy.go [hunk 4]
```diff
@@ -820,32 +820,52 @@ func (p *OAuthProxy) addHeadersForProxying(rw http.ResponseWriter, req *http.Req
...
+			req.Header["X-Forwarded-Access-Token"] = []string{session.AccessToken}
+		} else {
+			req.Header.Del("X-Auth-Request-Access-Token")
+		}
 	}
-	if p.PassAuthorization && session.IDToken != "" {
-		req.Header["Authorization"] = []string{fmt.Sprintf("Bearer %s", session.IDToken)}
+
+	if p.PassAuthorization {
+		if session.IDToken != "" {
```
Comment (steakunderscore) [inline oauthproxy.go]: Same here, no need to split the if statement.

File: oauthproxy.go [hunk 5]
```diff
@@ -820,32 +820,52 @@ func (p *OAuthProxy) addHeadersForProxying(rw http.ResponseWriter, req *http.Req
...
+			req.Header["Authorization"] = []string{fmt.Sprintf("Bearer %s", session.IDToken)}
+		} else {
+			req.Header.Del("Authorization")
+		}
 	}
-	if p.SetAuthorization && session.IDToken != "" {
-		rw.Header().Set("Authorization", fmt.Sprintf("Bearer %s", session.IDToken))
+	if p.SetAuthorization {
+		if session.IDToken != "" {
```
Comment (steakunderscore) [inline oauthproxy.go]: Same as above, no need to split the if statement.

File: oauthproxy.go [hunk 6]
```diff
@@ -820,32 +820,52 @@ func (p *OAuthProxy) addHeadersForProxying(rw http.ResponseWriter, req *http.Req
...
-			rw.Header().Set("X-Auth-Request-Access-Token", session.AccessToken)
+		} else {
+			rw.Header().Del("X-Auth-Request-Email")
 		}
 	}
-	if p.PassAccessToken && session.AccessToken != "" {
-		req.Header["X-Forwarded-Access-Token"] = []string{session.AccessToken}
+
+	if p.PassAccessToken {
+		if session.AccessToken != "" {
```
Comment (steakunderscore) [inline oauthproxy.go]: It looks like you have split the if conditions by mistake. Breaking the if statement in to two leaves a branch un handled.

File: oauthproxy.go [hunk 7]
```diff
@@ -820,32 +820,52 @@ func (p *OAuthProxy) addHeadersForProxying(rw http.ResponseWriter, req *http.Req
...
 		}
 	}
+
 	if p.SetXAuthRequest {
 		rw.Header().Set("X-Auth-Request-User", session.User)
 		if session.Email != "" {
 			rw.Header().Set("X-Auth-Request-Email", session.Email)
-		}
-		if p.PassAccessToken && session.AccessToken != "" {
-			rw.Header().Set("X-Auth-Request-Access-Token", session.AccessToken)
```
Comment (steakunderscore) [inline oauthproxy.go]: Why remove this setting `X-Auth-Request-Access-Token`?

File: oauthproxy.go (hunk 6 above)
Comment (ploxiln) [reply oauthproxy.go]: This way makes some sense because it always applies if you enable the `--pass-access-token` option. This option is something you have complete control over and does not vary with client state. If you don't enable this option, presumably your other applications do not expect this header to be set by `oauth2_proxy`, and you may prefer for oauth2_proxy to not mess with the parts of requests that do concern it. You might not be aware of all the headers that oauth2_proxy has options to set, and might be confused by some headers which you intended to pass through not being passed.

I'd also be fine with oauth2_proxy clearing all headers it could possibly be configured to set, not just those which its current configuration may set. There's not _too_ many optionally-enabled headers yet. I'm just explaining the reasoning for the way it was done here so far.

File: oauthproxy.go (hunk 7 above)
Comment (aeijdenberg) [reply oauthproxy.go]: I think I misread the previous code - it looked as though this was being set a few lines later, which is why I tried to remove the duplication, however from the looks of this diff I mucked it up, perhaps re-enforcing that the code is hard to follow.

I'll try to find time later this week (likely Friday) to correct - thanks for the pickup.

File: oauthproxy.go (hunk 6 above)
Comment (aeijdenberg) [reply oauthproxy.go]: The split is deliberate, for the reasons @ploxiln outlines - however setting different header *keys* in the next few lines is a copy/paste error. I'll do another pass.

File: oauthproxy.go (hunk 4 above)
Comment (aeijdenberg) [reply oauthproxy.go]: Disagree. The split is necessary so that an end user can't find a way to inject an ID token of their own to the backend that wasn't added by the proxy.

**Comment by aeijdenberg** [comment]
Rebased and pushed with fixes above.
